### PR TITLE
move out bootstrap token utils in versioned apis

### PR DESCRIPTION
**Pull request by dixudx**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Followup #77211

**Special notes for your reviewer**:
Depend on #77211

**Does this PR introduce a user-facing change?**:

```release-note
move out versioned bootstrap token utils to k8s.io/cluster-bootstrap
```

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
LGTM thanks @dixudx

**Comment by timothysc** [comment]
holding till the other PR is merged.

**Comment by dixudx** [comment]
Since v1beta2 is already out. We need to wait v1beta3 and make this changes in.

File: staging/src/k8s.io/cluster-bootstrap/util/tokens/tokens.go [hunk 1]
```diff
@@ -14,16 +14,21 @@ See the License for the specific language governing permissions and
...
-	bootstraputil "k8s.io/cluster-bootstrap/token/util"
+	"k8s.io/cluster-bootstrap/token/api"
+	legacyutil "k8s.io/cluster-bootstrap/token/util"
+)
+
+var (
+	bootstrapTokenRegexp = regexp.MustCompile(api.BootstrapTokenPattern)
 )
 
 // BootstrapTokenString is a token of the format abcdef.abcdef0123456789 that is used
```
Comment (rosti) [inline staging/src/k8s.io/cluster-bootstrap/util/tokens/tokens.go]: This has always looked to me as a versioned API and, indeed, it is part of the kubeadm versioned APIs. Perhaps we should introduce such an API as part of cluster-bootstrap?

@dixudx @neolit123 @fabriziopandini WDYT?

**Review by rosti — COMMENTED** [review]
Thanks for picking this up @dixudx !
So far it looks great, but I think, that we need to have the separated `BootstrapTokenString` type and its utility functions (in fact, interface implementations) to be versioned in `cluster-bootstrap` too. _That way we can import them in both v1beta1 and v1beta2._ (provided, that we can hide the implementation move for the released configs, but that may not be feasible)

File: staging/src/k8s.io/cluster-bootstrap/util/tokens/tokens.go (hunk 1 above)
Comment (neolit123) [reply staging/src/k8s.io/cluster-bootstrap/util/tokens/tokens.go]: yes, and we need to poke api-machinery about:
https://github.com/kubernetes/cluster-bootstrap#things-you-should-not-do
> Add API types to this repo. This is for the helpers, not for the types.

Comment (neolit123) [reply staging/src/k8s.io/cluster-bootstrap/util/tokens/tokens.go]: the alternative is to not have BoostrapTokenString as a public type under cluster-bootstrap, but rather as a private type with public helpers.

File: staging/src/k8s.io/cluster-bootstrap/util/secrets/secrets.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+		return ""
+	}
+	if val, ok := secret.Data[key]; ok {
+		return string(val)
+	}
+	return ""
+}
+
+// IsSecretExpired will identify whether the secret expires
+func IsSecretExpired(secret *v1.Secret, currentTime time.Time) bool {
```
[the remark this answers is not in the document]
Comment (dixudx) [reply staging/src/k8s.io/cluster-bootstrap/util/secrets/secrets.go]: Will do it when we v1beta3 is planned.

**Comment by dixudx** [comment]
> that we need to have the separated BootstrapTokenString type and its utility functions (in fact, interface implementations) to be versioned in cluster-bootstrap too. That way we can import them in both v1beta1 and v1beta2. (provided, that we can hide the implementation move for the released configs, but that may not be feasible)

@rosti Good suggestion. Will address this in v1beta3.

**Review by fabriziopandini — COMMENTED** [review]
@dixudx looks great!
+1 on @rosti suggestion to try to centralize BootstrapTokenString type as well

**Comment by neolit123** [comment]
we can have v1beta3 next release - 1.17.
BootstrapTokenString is a good enough reason for the new API version.

**Comment by rosti** [comment]
We first need to think where to place the extracted versioned APIs, complete with a new bootstrap token API group. Once we figure this out (write KEPs, implement them in there), we can start thinking of having a v1beta3 in kubeadm, that leverages on those.

**Comment by dixudx** [comment]
> We first need to think where to place the extracted versioned APIs, complete with a new bootstrap token API group.

k8s.io/cluster-bootstrap.

**Comment by neolit123** [comment]
@dixudx 
would you have time to perform this move for 1.17?
we are also adding v1beta3 in 1.17.

**Comment by dixudx** [comment]
> would you have time to perform this move for 1.17?
we are also adding v1beta3 in 1.17.

@neolit123 Sure. Will address this in the next few days.

**Review by fabriziopandini — COMMENTED** [review]
Thanks, @dixudx!
I'm ok with this change and I would like to get this in before starting the work on v1beta3. 
@rosti, @neolit123, opinions?

File: staging/src/k8s.io/cluster-bootstrap/util/tokens/tokens.go (hunk 1 above)
Comment (yastij) [reply staging/src/k8s.io/cluster-bootstrap/util/tokens/tokens.go]: it seems like https://github.com/kubernetes/cluster-bootstrap/blob/master/token/api landed last month (even though it contains only constants for now

File: staging/src/k8s.io/cluster-bootstrap/util/tokens/tokens.go [hunk 3]
```diff
@@ -36,3 +40,64 @@ func ParseToken(s string) (tokenID, tokenSecret string, err error) {
 	}
 	return split[1], split[2], nil
 }
+
+// BootstrapTokenString is a token of the format abcdef.abcdef0123456789 that is used
+// for both validation of the practically of the API server from a joining node's point
+// of view and as an authentication method for the node in the bootstrap phase of
+// "kubeadm join". This token is and should be short-lived
+type BootstrapTokenString struct {
```
Comment (yastij) [inline staging/src/k8s.io/cluster-bootstrap/util/tokens/tokens.go]: If we keep this, can we move it to https://github.com/kubernetes/cluster-bootstrap/blob/master/token/api ?

Comment (dixudx) [reply staging/src/k8s.io/cluster-bootstrap/util/tokens/tokens.go]: Current methods for `BootstrapTokenString` use several util functions. They are coupled. It is not a good pattern to import utils in api.

Comment (rosti) [reply staging/src/k8s.io/cluster-bootstrap/util/tokens/tokens.go]: In order to use `BootstrapTokenString` in a versioned API, it has to be versioned too. Therefore we need this, along with its interface implementation methods to be well versioned.
I can see the point, that we don't want to rely on much external stuff and I think, that we can safely avoid that while the types are versioned.

File: staging/src/k8s.io/cluster-bootstrap/util/tokens/tokens.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/cluster-bootstrap/util/tokens/tokens.go]: the only API is the type, structure, and helper methods to parse the bootstrap token secret contents, not the structures used to store decomposed bootstrap tokens in kubeadm config files

File: staging/src/k8s.io/cluster-bootstrap/util/tokens/tokens.go (hunk 3 above)
Comment (liggitt) [reply staging/src/k8s.io/cluster-bootstrap/util/tokens/tokens.go]: This is not a serializable struct (it does not contain json tags for serialization), and it is not versioned (it doesn't have a "hub" internal type with specific serialization versions to allow for different representations in the future).

File: cmd/kubeadm/app/apis/kubeadm/types.go [hunk 4]
```diff
@@ -19,6 +19,7 @@ package kubeadm
 import (
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	bootstraptokenutil "k8s.io/cluster-bootstrap/util/tokens"
```
Comment (liggitt) [inline cmd/kubeadm/app/apis/kubeadm/types.go]: the k8s.io/cluster-bootstrap/util/tokens package is not intended to define config APIs for decomposed bootstrap tokens

File: cmd/kubeadm/app/apis/kubeadm/v1beta2/types.go [hunk 5]
```diff
@@ -228,7 +229,7 @@ type Networking struct {
 type BootstrapToken struct {
 	// Token is used for establishing bidirectional trust between nodes and control-planes.
 	// Used for joining nodes in the cluster.
-	Token *BootstrapTokenString `json:"token"`
+	Token *bootstraptokenutil.BootstrapTokenString `json:"token"`
```
Comment (liggitt) [inline cmd/kubeadm/app/apis/kubeadm/v1beta2/types.go]: this is a breaking change, the bootstraptokenutil.BootstrapTokenString struct does not define json tags, so switching the serialized kubeadm type to use that struct would change serialization

**Review by liggitt — COMMENTED** [review]
The surface area of the bootstrap token API is limited to the type and structure of the Secret object and data keys. Ensuring the helpers to generate/parse those makes sense, but kubeadm's particular representation of the decomposed token for use in its own internal config file seems best left in the kubeadm config packages.

I am not in favor of moving these API types to the cluster-bootstrap package

**Comment by neolit123** [comment]
left related comment here:
https://github.com/kubernetes/enhancements/pull/1332#discussion_r338226491

**Comment by liggitt** [comment]
hoisting related comment from @neolit123 in https://github.com/kubernetes/enhancements/pull/1332#discussion_r338226491

> i believe @timothysc and @luxas originally had the idea to make the BoostrapToken(String) objects available outside of kubeadm.
> 
> https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/types.go#L247-L269
> 
> i'm not aware what prior discussions happened on this topic.
> 
> this is related to:
> https://github.com/kubernetes/kubernetes/pull/77669
> and
> https://github.com/kubernetes/kubeadm/issues/964
> 
> boostrap tokens being essentially a type of Secret (bootstrap.kubernetes.io/token) puts them in a bit of weird spot.
> 
> if the utilities here:
> https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/bootstraptokenhelpers.go
> don't get exposed outside of kubeadm, then it feels like a "boostrap token" will be an abstract type, due to the mostly undocumented "stringData" usage?
> 
> yet, we also have the option to:
> - close https://github.com/kubernetes/kubeadm/issues/964
> - close https://github.com/kubernetes/kubernetes/pull/77669
> - claim that the abstract type "bootstrap token" is GA because kubeadm, as the only publicly known consumer is GA and v1.Secret is GA.
> - document "stringData" for tokens
> - leave the token utils inside the kubeadm api under a seperate folder (cc @rosti, @dixudx )
>

**Comment by liggitt** [comment]
The bootstrap token authenticator lives at https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/auth/authenticator/token/bootstrap/bootstrap.go

The API surface it depends on are:
* kubernetes secrets
* the constants used as keys in the secret `data` map
* the structured parsing of the values in the secret `data` map
* the constant used as the secret type
* the secret name patterns/prefixes used to find the secret holding a particular token id

All of those are exposed in the following exported packages:
```
bootstrapapi "k8s.io/cluster-bootstrap/token/api"
bootstrapsecretutil "k8s.io/cluster-bootstrap/util/secrets"
bootstraptokenutil "k8s.io/cluster-bootstrap/util/tokens"
```

The bootstrap token config structure used by kubeadm is not part of the bootstrap token API, so I don't think it makes sense to move that to `k8s.io/cluster-bootstrap/token/api`

**Comment by neolit123** [comment]
> The bootstrap token config structure used by kubeadm is not part of the bootstrap token API, so I don't think it makes sense to move that to k8s.io/cluster-bootstrap/token/api

i think i'm starting to agree.
cc @timothysc do you have toughs?

also to this point:
> document "stringData" for tokens

i guess this is already covered here:
https://kubernetes.io/docs/reference/access-authn-authz/bootstrap-tokens/#bootstrap-token-secret-format

**Comment by timothysc** [comment]
Who consumes these types other then us?  Originally I was fine with them being helper utilities in client-go but daniel wanted them moved out which is why this PR exists.  I'm ok if we just put them in our types if that's what you're suggesting @neolit123 .

**Comment by timothysc** [comment]
> one question i have there is whether we should have them under bootstraptoken.kubeadm.k8s.io instead of kubeadm.k8s.io 

Off the cuff I kind of like that, but I'd want to think about it for a bit before we pull the trigger.

**Comment by rosti** [comment]
> one question i have there is whether we should have them under `bootstraptoken.kubeadm.k8s.io` instead of `kubeadm.k8s.io`

I like that idea. But we need to have a clear idea on how and what needs to be moved before execution.

**Comment by dixudx** [comment]
Below is the current struct,

```go
// BootstrapToken describes one bootstrap token, stored as a Secret in the cluster
type BootstrapToken struct {
	// Token is used for establishing bidirectional trust between nodes and control-planes.
	// Used for joining nodes in the cluster.
	Token *BootstrapTokenString `json:"token"`
	...
}

type BootstrapTokenString struct {
	ID     string `json:"-"`
	Secret string `json:"-"`
}
```

And I was wondering whether we could add a new filed (like `TokenStruct`, the filed name could be chosen a better one. This filed is only for inner usage, which will not be serialized/marshalled) to the struct here, like below

```go
type BootstrapToken struct {
	// Token is used for establishing bidirectional trust between nodes and control-planes.
	// Used for joining nodes in the cluster.
	Token *string `json:"token"`
        
        TokenStruct *BootstrapTokenString `json:"-"`
	...
}
```

An extra converting step is needed after current unmarshaling.  Or we could add such converting step when we calling nil `TokenTokenStruct`.

> observation these types are just wrappers to suite kubeadm.

By keeping wrapper `BootstrapTokenString` away from current marshaling, we could move out our `kubeadm/apis` to versioned staging apis.

**Pull request closed without merging**